src/xdp-utils.c

```diff
@@ -680,37 +706,18 @@ xdp_validate_serialized_icon (GVariant  *v,
       return FALSE;
     }
 
-  bytes = g_bytes_icon_get_bytes (G_BYTES_ICON (icon));
-  fd = g_file_open_tmp ("iconXXXXXX", &name, &error);
-  if (fd == -1)
+  bytes_fd = bytes_to_fd (g_bytes_icon_get_bytes (G_BYTES_ICON (icon)));
+  if (bytes_fd == -1)
     {
-      g_warning ("Icon validation: %s", error->message);
-      return FALSE;
-    }
-
-  stream = g_unix_output_stream_new (fd, FALSE);
-
-  /* Use write_all() instead of write_bytes() so we don't have to worry about
-   * partial writes (https://gitlab.gnome.org/GNOME/glib/-/issues/570).
-   */
-  bytes_data = g_bytes_get_data (bytes, &bytes_len);
-  if (!g_output_stream_write_all (stream, bytes_data, bytes_len, NULL, NULL, &error))
-    {
-      g_warning ("Icon validation: %s", error->message);
-      return FALSE;
-    }
-
-  if (!g_output_stream_close (stream, NULL, &error))
-    {
-      g_warning ("Icon validation: %s", error->message);
+      g_warning ("Icon validation: Rejecting icon because impossible to load bytes icon");
       return FALSE;
     }
 
   args[0] = icon_validator;
   args[1] = "--sandbox";
   args[2] = NULL;
 
-  key_file = spawn_validator(args, fd);
+  key_file = spawn_validator(args, bytes_fd);
```

Comment:
Alright, so this is kind of pointless. You're using the memfd as a source for the stdin of the validator process. What we really want is to pass the memfd fd to the validator process.

Reply:
Isn't stdin like any other fd in this case?

`GdkPixbufLoader`  only has a single way to load the image data via `gdk_pixbuf_loader_write()` so it shouldn't matter if we pass a fd or map the stdin to the memfd directly.

Reply:
sure, the problem is that stdin is a pipe (or rather a thing you call read on) so you're copying memory from the memfd to a pipe and then the receiving process copies that from the pipe back to somewhere else. What we want is to mmap the memfd to avoid any copies.

Technically you could pass the memfd directly as stdin and mmap that. It's just not the convention and the code in this PR doesn't do it.

Reply:
Also, when you mmap the memfd, you can pass the entire thing to `gdk_pixbuf_loader_write` in a single call.